### PR TITLE
New Keyboard: Wolfmark Club WM1

### PR DESCRIPTION
**Pull request by mechmerlin**
## Description

With much help from @tzarc we finally have QMK working on this guy! Thanks to mrpetrov for lending me his board to get this done on. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/wolfmarkclub/wm1/readme.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Wolfmark Club WM1
+
+![%KEYBOARD%](https://i.imgur.com/TDFp97y.jpg)
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/readme.md:3]:
```suggestion
![WM1](https://i.imgur.com/TDFp97y.jpg)
```

File: keyboards/wolfmarkclub/wm1/readme.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+Split Ergo 65% keyboard
+
+* Keyboard Maintainer: [MechMerlin](https://github.com/mechmerlin)
+* Hardware Supported: WM1
+* Hardware Availability:
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/readme.md:9]: ?

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Build Options
+#   change yes to no to disable
+#
+
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/rules.mk:4]:
```suggestion
```

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support (+2400 to 4200, depending on config)
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs (+400)
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/rules.mk:36]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = yes       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
MIDI_ENABLE = no            # MIDI support
UNICODE_ENABLE = no         # Unicode
BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
AUDIO_ENABLE = no           # Audio output on port C6
FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
```

File: keyboards/wolfmarkclub/wm1/wm1.c [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    matrix_scan_user();
+}
+
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+    // put your per-action keyboard code here
+    // runs for every action, just before processing by the firmware
+
+    return process_record_user(keycode, record);
+}
+
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/wm1.c:26]:
```suggestion
```

File: keyboards/wolfmarkclub/wm1/wm1.c [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#include "wm1.h"
+
+void matrix_init_kb(void) {
+    // put your keyboard start-up code here
+    // runs once when the firmware starts up
+
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/wm1.c:6]:
```suggestion
```

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+MCU_LDSCRIPT = wm1_f103
+BOARD = STM32_F103_STM32DUINO
+
+DFU_ARGS = -d 1eaf:0003 -a2 -R
+DFU_SUFFIX_ARGS = -v 1eaf -p 0003
+
+
+# Enter lower-power sleep mode when on the ChibiOS idle thread
+OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
+
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/rules.mk:19]:
```suggestion
```
Moved this to the bottom.

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+NKRO_ENABLE = yes            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support (+2400 to 4200, depending on config)
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs (+400)
+LTO_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/rules.mk:37]:
```suggestion
LTO_ENABLE = yes

# Enter lower-power sleep mode when on the ChibiOS idle thread
OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
```

File: keyboards/wolfmarkclub/wm1/config.h [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x6465
```
Comment (fauxpark) [inline keyboards/wolfmarkclub/wm1/config.h:22]:
```suggestion
#define PRODUCT_ID      0x2B29
```
```
echo -n "WM1" | shasum
```

File: keyboards/wolfmarkclub/wm1/bootloader_defs.h [hunk 10]
```diff
@@ -0,0 +1,10 @@
+/* Address for jumping to bootloader on STM32 chips. */
+/* It is chip dependent, the correct number can be looked up here (page 175):
+ * http://www.st.com/web/en/resource/technical/document/application_note/CD00167594.pdf
+ * This also requires a patch to chibios:
+ *  <tmk_dir>/tmk_core/tool/chibios/ch-bootloader-jump.patch
+ */
+
+// STM32F103* does NOT have an USB bootloader in ROM (only serial),
+//  so setting anything here does not make much sense
+#define STM32_BOOTLOADER_ADDRESS 0x80000000
```
Comment (tzarc) [inline keyboards/wolfmarkclub/wm1/bootloader_defs.h:10]:
```suggestion
#define STM32_BOOTLOADER_ADDRESS 0x08000000
```

Can we make this change, but also do a quick test to add `RESET` to a keymap to see if it'll actually enter bootloader? The stock bootloader will be at `0x08000000` so if we set it up that we perhaps we can ensure the keycode works too.

File: keyboards/wolfmarkclub/wm1/ld/wm1_f103.ld [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+/*
+ * ST32F103xB memory setup for use with the maplemini bootloader.
+ * You will have to
+ * 	#define CORTEX_VTOR_INIT 0x5000
+ * in your projects chconf.h
+ */
```
Comment (tzarc) [inline keyboards/wolfmarkclub/wm1/ld/wm1_f103.ld:22]:
```suggestion
/*
 * STM32F103xC memory setup for use with the WM1 mass-storage device bootloader.
 * You will have to change:
 * 	OPT_DEFS = -DCORTEX_VTOR_INIT=0x10000
 * in your board's rules.mk
 */
```

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+MCU = STM32F103
+
+# GENERIC STM32F103C8T6 board - stm32duino bootloader
+OPT_DEFS = -DCORTEX_VTOR_INIT=0x10000
+MCU_LDSCRIPT = wm1_f103
+BOARD = STM32_F103_STM32DUINO
+
+DFU_ARGS = -d 1eaf:0003 -a2 -R
+DFU_SUFFIX_ARGS = -v 1eaf -p 0003
+
```
Comment (tzarc) [inline keyboards/wolfmarkclub/wm1/rules.mk:15]:
```suggestion
```

File: keyboards/wolfmarkclub/wm1/bootloader_defs.h [hunk 13]
```diff
@@ -0,0 +1,10 @@
+/* Address for jumping to bootloader on STM32 chips. */
+/* It is chip dependent, the correct number can be looked up here (page 175):
+ * http://www.st.com/web/en/resource/technical/document/application_note/CD00167594.pdf
+ * This also requires a patch to chibios:
+ *  <tmk_dir>/tmk_core/tool/chibios/ch-bootloader-jump.patch
+ */
+
+// STM32F103* does NOT have an USB bootloader in ROM (only serial),
+//  so setting anything here does not make much sense
```
Comment (tzarc) [inline keyboards/wolfmarkclub/wm1/bootloader_defs.h:9]:
```suggestion
STM32F103 does not have a USB bootloader in ROM -- the WM1 has a usermode bootloader at the normal boot address -- `0x08000000`.
```

File: keyboards/wolfmarkclub/wm1/config.h [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+#define RGBLIGHT_SLEEP  /* If defined, the RGB lighting will be switched off when the host goes to sleep */
+#endif
+/* define if matrix has ghost */
+//#define MATRIX_HAS_GHOST
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCE    5
+
+#define TAPPING_TERM 500
+
```
Comment (zvecr) [inline keyboards/wolfmarkclub/wm1/config.h:61]:
```suggestion

```

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 15]
```diff
@@ -0,0 +1,37 @@
+# Build Options
+#   change yes to no to disable
+#
+
+# MCU name
+MCU = STM32F103
+
+# GENERIC STM32F103C8T6 board - stm32duino bootloader
```
Comment (zvecr) [inline keyboards/wolfmarkclub/wm1/rules.mk:8]:
```suggestion
# GENERIC STM32F103C8T6 board - mass storage bootloader
```

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+OPT_DEFS = -DCORTEX_VTOR_INIT=0x10000
+MCU_LDSCRIPT = wm1_f103
+BOARD = STM32_F103_STM32DUINO
+
+DFU_ARGS = -d 1eaf:0003 -a2 -R
+DFU_SUFFIX_ARGS = -v 1eaf -p 0003
+
+
```
Comment (zvecr) [inline keyboards/wolfmarkclub/wm1/rules.mk:16]: Stop `:flash` from doing stuff
```suggestion

PROGRAM_CMD=echo 'CLI flashing not supported' >&2

```

File: keyboards/wolfmarkclub/wm1/rules.mk (hunk 15 above)
Comment (tzarc) [reply keyboards/wolfmarkclub/wm1/rules.mk:8]: it's an RCT6, I guess we should do a full pass

File: keyboards/wolfmarkclub/wm1/bootloader_defs.h (hunk 10 above)
Comment (mechmerlin) [reply keyboards/wolfmarkclub/wm1/bootloader_defs.h:11]: I will ignore your comment

Comment (tzarc) [reply keyboards/wolfmarkclub/wm1/bootloader_defs.h:11]: Lets just ignore this for now, unless you're holding down ESC during bootup the bootloader will end up executing QMK anyway.

**Comment by yulei** [comment]
the original codes were here:)

https://github.com/yulei/qmk_firmware/blob/wm690/keyboards/wm1/v3/

The codes should be broken with the latest qmk, but the pin definitions could be an reference source.

**Comment by mechmerlin** [comment]
> the original codes were here:)
> 
> https://github.com/yulei/qmk_firmware/blob/wm690/keyboards/wm1/v3/
> 
> The codes should be broken with the latest qmk, but the pin definitions could be an reference source.

oof, this would have been super helpful lol

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = yes        # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = yes       # Breathing sleep LED during USB suspend
```
Comment (noroadsleft) [inline keyboards/wolfmarkclub/wm1/rules.mk:20]:
```suggestion
SLEEP_LED_ENABLE = yes      # Breathing sleep LED during USB suspend
```

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover
```
Comment (noroadsleft) [inline keyboards/wolfmarkclub/wm1/rules.mk:22]:
```suggestion
NKRO_ENABLE = yes           # USB Nkey Rollover
```

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
```
Comment (noroadsleft) [inline keyboards/wolfmarkclub/wm1/rules.mk:24]:
```suggestion
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
```

File: keyboards/wolfmarkclub/wm1/rules.mk [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
```
Comment (noroadsleft) [inline keyboards/wolfmarkclub/wm1/rules.mk:30]:
```suggestion
```

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**